clang/lib/AST/PreorderAST.cpp

```diff
@@ -29,24 +29,30 @@ void PreorderAST::AddNode(Node *N, Node *Parent) {
   }
 }
 
-void PreorderAST::CoalesceNode(BinaryNode *B, BinaryNode *Parent) {
+void PreorderAST::RemoveNode(Node *N, Node *Parent) {
+  assert(isa<BinaryNode>(Parent) && "Invalid parent");
+
+  auto *P = dyn_cast<BinaryNode>(Parent);
+
   // Remove the current node from the list of children of its parent.
-  for (auto I = Parent->Children.begin(),
-            E = Parent->Children.end(); I != E; ++I) {
-    if (*I == B) {
-      Parent->Children.erase(I);
+  for (auto I = P->Children.begin(),
+            E = P->Children.end(); I != E; ++I) {
+    if (*I == N) {
+      P->Children.erase(I);
       break;
     }
   }
 
-  // Move all children of the current node to its parent.
-  for (auto *Child : B->Children) {
-    Child->Parent = Parent;
-    Parent->Children.push_back(Child);
+  if (auto *B = dyn_cast<BinaryNode>(N)) {
```

Comment:
This is only correct if (1) the child is singleton node (2) the parent and the current node are associative and commutative operators, and are the same operator.  It would be good to have an assert to prevent misuse of this function.

Reply:
I am not sure I understand your comment. Basically, if N is a BinaryNode we want to move its children to the parent of N and then delete N. If not, then we can simply delete N.